### DISCUSSION TITLE
Content Collections Discussion

### DISCUSSION THREAD

**Discussion by louiss0** [withastro/roadmap#415] [Proposal]
# Content Collections Criticism

First off Id like to say a good job on this RFC. It is good enough for me to create a project around and good enough for me to maybe scale it. I like the fact that you have decided to put the render function on the entry instead of us having to import it. I like get `collectionToPaths()` `function`  **Please add it!** I even like the fact that Zod was chosen for this RFC. You said that maybe in the future other formats could be supported. I hope the next one is JSON. But no tool is perfect. This is only the beginning so I have decided to talk about some big changes to consider.                  

## Magic Properties

In Astro there is so far two magic frontmatter properties that are available. 
The first one is `draft:` and the second one is `layout`. I believe that both of them should be removed since people are expected to create their blogs by using Content Collections. Or they should at least not be able to be used in the `/content` folder. The reason why is that these properties are just not needed anymore. If the layout property was to be used it would lead to bad design and coupling. When you use the layout key the render function activates the mechanism responsible for rendering layouts. This RFC is better of being used to make the developer have to import the layout that the person wants to use inside of the page he or she wants to use it. Having magic layouts in the `/content` folder can only lead to bad behavior.      

I could argue about removing the magic properties completely but then when it comes to pages the user would have to write lots of boilerplate code inside of pages. Since the `/content` folder exists the only things `.mdx` is good for is just being used as a templating language and importing components. So I think it's just better to just remove them from the content folder. Now as far as `draft:` is concerned. I'd argue that if someone was thinking about whether or not a page should be a draft of not it should just be put in the content folder instead. But draft is a minor thing. But if it is going to exist as a first-class citizen I'd suggest letting the developer know about it throughout the creation of each collection through types.    

## Extending and Defining a default Schema 

There is a problem with the schema property of define collection. It only allows me to define a `ZodRawShape` I can't use `z.object()` on it at all. This means I can't use other features of `zod` in order to construct my schemas at all. This is not good. This means that if a developer wants to be able to reuse other schema definitions like **title author draft `updatedDate` and even `pubDate`** they would have to rewrite them all over again. Remember the **DRY** principle.       

```js:
import { z, defineCollection } from 'astro:content';

// In this example the title is repeated twice 
const releases = defineCollection({
  schema: {
    title: z.string(), // here
    version: z.number(),
  },
});

const engineeringBlog = defineCollection({
  schema: {
    title: z.string(), //and  here
    tags: z.array(z.string()),
    image: z.string().optional(),
  },
});

export const collections = {
  releases: releases,
  // Don't forget 'quotes' for collection names containing dashes
  'engineering-blog': engineeringBlog,
};
```

**I think there should be a default schema for people to use.  I'd put it in the astro config but it could also exist in the `/config.ts` file for the content folder.** 

```js
{
 contentCollections: {
	 defaultSchema: {
		 title: z.string().max(90)
		 author: z.string("Authors name").default("Authors name")
	 } 
 }
}

```

**Users would probably then have the power to extend it by importing it from `astro:content`**

```js

import {defaultSchema, z } from "astro:content"

export const collections = {
	blog: defineCollection({
		schema: defaultSchema 
	})
}

```

I'm asking for the developer to have more access to `zod`'s features. The schema key expects just a plain object.  That makes it so that I can't just use the other functions that zod provides.  If you don't intend to let developers gain full access to the API  of zod through define collection. You could at least give them back some of its capabilities by using the `extends:` key in define collections and the `omit:` key so that you can omit some keys from a schema. 

## Schema for injected frontmatter 

At the moment injected frontmatter is just not typeable at all. I think the user should have access to all the types for the frontmatter that is injected into the pages. I would like to have an `injectedFrontmatterSchema:` available for collections. If that is not possible I the `render()` function should have a generic argument that allows the user to pass in a type.

**Injected Frontmatter Schema**

```js

const blog = defineCollection({
injectedFrontmatterSchema:{
 readingTime: z.string().datetime(),
 author: z.string().default("Shelton Louis"),
 
	} 
})

```

**Example with the render function**

```js

render<T extends Record<string, unknown> >(): Promise<{  
Content: AstroComponentFactory;  
headings: MarkdownHeading[];  
injectedFrontmatter: T;  
}>

```

## Lastly

I don't have many other concerns from here but are going to find a better way of generating types from collections. Generating an entry map for each individual collection seems good in the short run but bad in the long run.  The thing is that the map can become huge and `ts` may not be able to tell us the answers we need and there could be some scalability issues when it comes to writing and erasing types. 

I wish there was a way to specify injected front matter as default and each collection as well. That way people don't have to keep having to read the file where they put remark plugins to find which ones they injected. A key to specify injected front Matter Schemas would be nice.

**Comment by aFuzzyBear** [comment]
Hey @louiss0 thank you for the feedback, I agree this is very early days and the scope for the `Content Collection API` was deliberately kept small as it was an exercise in its own right. These are good thoughts that you have outlines. I know extending `zod` would be something that would would like to slowly pursue in the future, but in due course as we are really introducing something very new and some can consider an advance concept to our users. It lets us be a bit more gentle with our approach. 
These are great ideas, and I know that we are only in the beginning of seeing where this set of APIs can take our users, 
💚🤘
cc: @bholmesdev

**Reply by louiss0** [reply]
I was also hoping for a function that allows you to get multiple collections as well. I'd call it `getMultipleCollections()`. I'm going to have to visit another RFC for Astro that I wrote. It has something to do with the Astro island component. I think it's good to just write a new RFC instead. I hope you remember what was said here today and thanks for taking the time to read this.

**Comment by bholmesdev** [comment]
Thanks for raising these points Louis! A lot of these ideas overlap with existing plans :)
- We agree that content collections make "magic" properties like `layout` less necessary by leaning on dynamic routes. To ensure migration is smooth, we felt maintaining existing magic would offer the best experience. Still, it may be worth documenting how `layout` and `draft` interact with content collections. We do plan to revisit these properties as we approach a v2.
- Yes, we've discussed default schemas and CLI wizards as part of Astro core. We would likely provide schemas for common use cases based on user feedback, and how we observe users applying content collections. A default Blog schema feels like low-hanging fruit. If you want to bundle some defaults of your own into an npm package, I'd love to check it out.
- We're aware that `injectedFrontmatter` feels a bit tacked-on. However, separating schemas from the remark pipeline proved vital for performance gains, so we considered this out-of-scope for the initial RFC. For strong type checking, I expect us to find a different solution to injected frontmatter entirely. We will share ideas as publicly as possible once we prototype this.

**Reply by louiss0** [reply]
I do hope this RFC becomes powerful by the end of version two. I hope it is supported with a good query mechanism. Filtering is good but not enough. I'm glad you will think about my criticisms and I also want a function that looks for multiple collections as well. I think this is it for now I can do what I need to do with the tools you provided for me.

**Comment by Temez1** [comment]
I like that Astro is starting to think about content more seriously, this is a great direction and good work so far! 

I even had a short [discussion ](https://github.com/withastro/rfcs/discussions/334)about this, when I was researching Astro. In short

> It would be good for integrations to be able to separate the content away from source code. Eg. In my case I'm planning to integrate Stackbit later to the project that mentions
> — "Stackbit's only real requirement is that content must be structured and separated from code."

Having said that, I noted that for now the content directory won't support other than md(x) files.

> Files other than your .md/.mdx content are not permitted inside src/content/

I hope the content folder could be a place for all the content in the future because of integrations and clean content/code relationship. The [Contentlayer](https://www.contentlayer.dev/) project could be a good part as it solves many of the same problems I believe part of this is trying to solve.

A minor detail would be, that the content directory could reside outside `src/` and the content would work with optimized images etc. I'm not aware of the obstacles for this.

**Reply by bholmesdev** [reply]
Hey, glad to hear you're liking it @Temez1! Too address those comments:
- Yes, supporting more data types like `.json`, `.yaml`, or `.csv` are top-of-mind. We stuck with generic terminology like "collection" and "data" for frontmatter to keep this door open!
- Yes, ContentLayer is a notable inspiration for where content collections could go. I'm curious what needs you have right now that content collections could address better. Is integration with Stackbit something on your mind? And what could that integration look like?
- Making the location of your content directory configurable has come up a few times. Do you have a limitation to using `src/content/` today?

**Reply by Temez1** [reply]
Hey @bholmesdev!

Thanks for the quick response and good questions.

I'm not yet implementing content related things, just researching and raising the concerns as implementing this is only matter of time when projects get more complex. But yeah, integration of Stackbit is one of the possibilities depending if&what headless CMS I would integrate. For now I will stick to file based CMS to get started.

I don't have any limitations for the location of content directory and this is just minor preference. And I'm not aware that there would be any limitations in the future.

For simplicity and separation of content and code, I would imagine the root to be like 

```
content/
- images
- pages
- ..

src/
- ..
```
In the end it might be that I stick with git based CMS and will use the Content folder if the projects stay simple enough :) 

My major concern is how to i18n the content in a future proof way. Feels like I'm running in circles between headless CMS, the i18next astro integration & how these would fit to this. Overthinking to future proof as usual 😅 However, this would be the first core functional for almost any english as second language county website like my project is.

There isn't official documentation or recommendations from Astro, but do you have any suggestions using this how should I add support for languages? I saw the docs mentioned that each would have own directory with related content, but in practice I feel there's a lot more to consider and would appreciate any tips. 🙏 Thanks✌️

**Reply by naiyerasif** [reply]
@bholmesdev It would be great to have more control over the location of `src/content` because in many cases we don't have any say over the location of the content. It may not be helpful for folks starting from scratch but it'd be great for those who want to migrate from other frameworks.

**Comment by louiss0** [comment]
I have recently realized two things but did not put them here. First there is no get entry type. Second there is no `getCollectionData()`. I think these two methods are important. Since you don't give use the full power of Zod instead you need to have a custom way of deriving types from schemas There is no good way for the developer to just use the type of the schema coming from Zod. The type I want created is called the `ExtractData`. I want the `getEntryData` along with `getCollectionData` methods as well. 

**The type Extract Data should look like this**.

```ts

type ExtractData<C extends keyof typeof entryMap> = CollectionEntry<C>['data']

```

Both `getCollectionData` and `getEntryData` should just return data. You could add a third option to the `getEntry` and the `getCollection` methods in order to just get data 


```js
// third parameter to just get the data for both functions
getCollection("project", null, true)

getEntry("project", 'post.md', true)

```
